odl/space/ntuples.py

```diff
@@ -1767,55 +1939,103 @@ def inner(self, x1, x2):
         inner : float or complex
             The inner product of the two provided vectors
         """
-        inner = _inner_default(x1 * self.vector, x2)
-        if is_real_dtype(x1.dtype):
-            return float(inner)
+        if self.exponent != 2.0:
+            raise NotImplementedError('No inner product defined for '
+                                      'exponent != 2 (got {}).'
+                                      ''.format(self.exponent))
+        else:
+            inner = _inner_default(x1 * self.vector, x2)
+            if is_real_dtype(x1.dtype):
+                return float(inner)
+            else:
+                return complex(inner)
+
+    def norm(self, x):
+        """Calculate the vector-weighted norm of a vector.
+
+        Parameters
+        ----------
+        x : `Fn.Vector`
+            Vector whose norm is calculated
+
+        Returns
+        -------
+        norm : float
+            The norm of the provided vector
+        """
+        if self.exponent == 2.0:
+            return super().norm(x)
```

Comment:
What I mean is that currently `_FnWeightingBase` implements `norm` using `inner`. To me this isnt obviously the way to go, for example, one might expect the base one to be optimized properly. In the `norm` case this doesnt realy matter, but currently `dist` requires a copy. What if we decided to optimize it in `_FnWeightingBase`? Then this code would break (in the `dist` case).

I would in this case assume that `super().norm()` gave me the "standard R^n norm", not some magic depending on how `inner` is implemented in this function.